rust/theoros/src/services/hyperlane/mod.rs

```diff
@@ -30,59 +32,153 @@ impl Service for HyperlaneService {
 }
 
 impl HyperlaneService {
-    pub fn new(state: AppState) -> Self {
-        Self { state }
+    pub fn new(storage: Arc<TheorosStorage>) -> Self {
+        Self { storage }
     }
 
-    /// Every [FETCH_INTERVAL] seconds, fetch the latest checkpoint signed for all
-    /// registered validators.
     pub async fn run_forever(&self) -> anyhow::Result<()> {
         loop {
             self.process_validator_checkpoints().await;
             tokio::time::sleep(FETCH_INTERVAL).await;
         }
     }
 
+    /// Processes validator checkpoints by fetching signed checkpoints from all validators for each unsigned nonce.
+    ///
+    /// This function performs the following steps:
+    ///
+    /// 1. **Retrieve Unsigned Nonces**:
+    ///    - Fetches all the nonces currently stored in the `UnsignedCheckpointsStorage`.
+    ///
+    /// 2. **Retrieve Validators and Fetchers**:
+    ///    - Gets all registered validators and their corresponding fetchers from the `ValidatorsFetchersStorage`.
+    ///
+    /// 3. **Fetch Signed Checkpoints**:
+    ///    - Attempts to fetch the signed checkpoint all unsigned nonce from each validator's fetcher (in parallel),
+    ///
+    /// 4. **Process Completed Nonces**:
+    ///    - After all fetches are completed, iterates over the unsigned nonces again.
+    ///    - Checks if all validators have signed the nonce using the `all_validators_signed_nonce` method.
+    ///    - **Note**: Currently, the function only proceeds if **all** validators have signed the nonce.
+    ///      - There's a `TODO` to modify this behavior to use a quorum method (e.g., consider a nonce as valid if 66% of validators have signed it).
+    ///    - If all validators have signed the nonce:
+    ///        - Calls `store_event_updates(nonce)` to process and store the updates associated with that nonce.
+    ///        - Removes the nonce from the `UnsignedCheckpointsStorage`, as it has been fully processed.
+    ///
     async fn process_validator_checkpoints(&self) {
-        let validators_locations = self.state.storage.validators_locations().all().await;
-        let futures = validators_locations
-            .into_iter()
-            .map(|(validator, fetcher)| self.process_single_validator(validator, fetcher));
+        let unsigned_nonces = self.storage.unsigned_checkpoints().nonces().await;
+        if unsigned_nonces.is_empty() {
+            return;
+        }
 
+        let validators_fetchers = self.storage.validators_fetchers().all().await;
+        let mut futures = Vec::with_capacity(unsigned_nonces.len());
+        for &nonce in &unsigned_nonces {
+            for (validator, fetcher) in &validators_fetchers {
+                let fut = self.fetch_checkpoint_for_validator(*validator, fetcher.clone(), nonce);
+                futures.push(fut);
+            }
+        }
         futures::future::join_all(futures).await;
```

Comment:
we don't propagate errors here ? It's fine since no errors are returned anyway but then why return a result in the first place ?

Reply:
ah true, removed